hw/mcu/stm/stm32_common/src/hal_nvreg.c

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <mcu/cortex_m4.h>
+#include <hal/hal_nvreg.h>
+#include <mcu/stm32_hal.h>
+
+/* Use header files to determine number of RTC Backup registers */
+#if defined RTC_BKP31R
+/* Some devices that have RTC backup registers have 32 */
+#define HAL_NVREG_MAX (32)
+#elif defined RTC_BKP15R
+/* Some devices that have RTC backup registers have 16 */
+#define HAL_NVREG_MAX (16)
+#elif defined RTC_BKP4R
+/* All devices that have RTC backup registers have at least 5 */
+#define HAL_NVREG_MAX (5)
+#else
+/* Some devices don't have any backup registers */
+#define HAL_NVREG_MAX (0)
+#endif
+
+/* RTC backup registers are 32-bits wide */
+#define HAL_NVREG_WIDTH_BYTES (4)
+
+static volatile uint32_t *regs[HAL_NVREG_MAX] = {
+#if HAL_NVREG_MAX > 0
+    &RTC->BKP0R,
+    &RTC->BKP1R,
+    &RTC->BKP2R,
+    &RTC->BKP3R,
+    &RTC->BKP4R,
+#if HAL_NVREG_MAX > 5
+    &RTC->BKP5R,
+    &RTC->BKP6R,
+    &RTC->BKP7R,
+    &RTC->BKP8R,
+    &RTC->BKP9R,
+    &RTC->BKP10R,
+    &RTC->BKP11R,
+    &RTC->BKP12R,
+    &RTC->BKP13R,
+    &RTC->BKP14R,
+    &RTC->BKP15R,
+#if HAL_NVREG_MAX > 16
+    &RTC->BKP16R,
+    &RTC->BKP17R,
+    &RTC->BKP18R,
+    &RTC->BKP19R,
+    &RTC->BKP20R,
```

Comment:
You can add `#if HAL_NVREG_MAX > 20` because many devices have that amount of registers.

Reply:
Changed it to match what they themselves have in the rtc header files.
For example: https://github.com/apache/mynewt-core/blob/master/hw/mcu/stm/stm32l4xx/src/ext/Drivers/STM32L4xx_HAL_Driver/Inc/stm32l4xx_ll_rtc.h#L562